File: GddsClient.java
package com.gddsstreaming.client;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import java.util.concurrent.CountDownLatch;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.util.Scanner;

public class GddsClient {
    private final GddsStreamServiceGrpc.GddsStreamServiceStub asyncStub;
    private final GddsStreamServiceGrpc.GddsStreamServiceBlockingStub blockingStub;
    private String clientId;
    private final CountDownLatch latch = new CountDownLatch(1);

    public enum VehicleType {
        UNKNOWN(0),
        CAR(1),
        TRUCK(2),
        BUS(3),
        MOTORCYCLE(4);

        private final int value;

        VehicleType(int value) {
            this.value = value;
        }

        public static VehicleType valueOf(int value) {
            for (VehicleType type : values()) {
                if (type.value == value) {
                    return type;
                }
            }
            return UNKNOWN;
        }
    }

    public GddsClient(String serverAddress) {
        if (serverAddress == null || serverAddress.isEmpty()) {
            throw new IllegalArgumentException("Server address is required");
        }

        // Create gRPC channel
        ManagedChannel channel = ManagedChannelBuilder.forTarget(serverAddress)
                .usePlaintext() // Use plaintext for simplicity; configure SSL for production
                .build();

        // Create stubs for async and blocking calls
        asyncStub = GddsStreamServiceGrpc.newStub(channel);
        blockingStub = GddsStreamServiceGrpc.newBlockingStub(channel);
    }

    public String getClientId() {
        try {
            GenerateClientIdRequest request = GenerateClientIdRequest.newBuilder().build();
            GenerateClientIdResponse response = blockingStub.generateClientId(request);
            clientId = response.getClientId();
            return clientId;
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate client ID: " + e.getMessage(), e);
        }
    }

    public void startReceiving() {
        if (clientId == null || clientId.isEmpty()) {
            throw new IllegalStateException("Client ID not set. Call getClientId first.");
        }

        SubscriptionRequest request = SubscriptionRequest.newBuilder()
                .setClientId(clientId)
                .build();

        asyncStub.subscribe(request, new StreamObserver<GddsMessage>() {
            @Override
            public void onNext(GddsMessage gddsMessage) {
                System.out.printf("Received GDDS: IP=%s, Vehicle=%s, Lat=%f, Lon=%f, Speed=%f, " +
                                "Bearing=%f, TransitMode=%s, Timestamp=%s, ServerTimestamp=%s, Message=%s%n",
                        gddsMessage.getIp(),
                        gddsMessage.getVeh(),
                        gddsMessage.getLat(),
                        gddsMessage.getLon(),
                        gddsMessage.getSpd(),
                        gddsMessage.getHdg(),
                        VehicleType.valueOf(gddsMessage.getTrnTyp()),
                        OffsetDateTime.ofInstant(Instant.ofEpochMilli(gddsMessage.getTm()), java.time.ZoneOffset.UTC),
                        OffsetDateTime.ofInstant(Instant.ofEpochMilli(gddsMessage.getSvrtm()), java.time.ZoneOffset.UTC),
                        gddsMessage.getMsg());
            }

            @Override
            public void onError(Throwable t) {
                if (t instanceof io.grpc.StatusRuntimeException) {
                    io.grpc.Status status = ((io.grpc.StatusRuntimeException) t).getStatus();
                    if (status.getCode() == Status.Code.CANCELLED) {
                        System.out.println("Client " + clientId + " subscription was cancelled.");
                    } else if (status.getCode() == Status.Code.UNAUTHENTICATED) {
                        System.out.println("Client " + clientId + " failed to subscribe: Invalid or unauthorized client ID.");
                    } else {
                        System.out.println("Error in subscription for client " + clientId + ": " + t.getMessage());
                    }
                } else {
                    System.out.println("Error in subscription for client " + clientId + ": " + t.getMessage());
                }
                latch.countDown();
            }

            @Override
            public void onCompleted() {
                System.out.println("Client " + clientId + " stream completed.");
                latch.countDown();
            }
        });

        System.out.println("Client " + clientId + " subscribed to GDDS stream.");
    }

    public void terminate() {
        if (clientId == null || clientId.isEmpty()) {
            System.out.println("No client ID set. Nothing to terminate.");
            return;
        }

        try {
            TerminationRequest request = TerminationRequest.newBuilder()
                    .setClientId(clientId)
                    .build();
            TerminationResponse response = blockingStub.terminate(request);

            if (response.getSuccess()) {
                System.out.println("Client " + clientId + " successfully terminated subscription.");
            } else {
                System.out.println("Client " + clientId + " termination failed.");
            }
        } catch (Exception e) {
            System.out.println("Error terminating client " + clientId + ": " + e.getMessage());
        } finally {
            latch.countDown();
        }
    }

    public void awaitTermination() {
        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.out.println("Await termination interrupted: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        String serverAddress = "localhost:50051"; // Replace with your server address

        GddsClient client = new GddsClient(serverAddress);

        // Request a client ID
        String clientId = client.getClientId();
        System.out.println("Received client ID: " + clientId);

        // Start receiving in a separate thread
        new Thread(client::startReceiving).start();

        // Wait for user input to terminate
        System.out.println("Press Enter to terminate the subscription...");
        new Scanner(System.in).nextLine();

        // Terminate the subscription
        client.terminate();

        // Wait for the stream to complete
        client.awaitTermination();

        System.out.println("Client shut down.");
    }
}

File: GddsClient.cs
using Grpc.Core;
using Grpc.Net.Client;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace GddsStreaming.Client
{
    public class GddsClient
    {
        private readonly GddsStreamService.GddsStreamServiceClient _client;
        private readonly string _serverAddress;
        private string? _clientId;
        private readonly CancellationTokenSource _cancellationTokenSource;

        public GddsClient(string serverAddress)
        {
            if (string.IsNullOrEmpty(serverAddress))
                throw new ArgumentException("Server address is required", nameof(serverAddress));

            _serverAddress = serverAddress;
            _cancellationTokenSource = new CancellationTokenSource();

            // Create gRPC channel and client
            var channel = GrpcChannel.ForAddress(serverAddress, new GrpcChannelOptions
            {
                // Optional: Configure channel options (e.g., SSL, credentials)
            });
            _client = new GddsStreamService.GddsStreamServiceClient(channel);
        }

        public async Task<string> GetClientIdAsync()
        {
            try
            {
                var response = await _client.GenerateClientIdAsync(new GenerateClientIdRequest());
                _clientId = response.ClientId;
                return _clientId;
            }
            catch (Exception ex)
            {
                throw new Exception($"Failed to generate client ID: {ex.Message}", ex);
            }
        }

        public async Task StartReceivingAsync()
        {
            if (string.IsNullOrEmpty(_clientId))
            {
                throw new InvalidOperationException("Client ID not set. Call GetClientIdAsync first.");
            }

            try
            {
                // Subscribe to the GDDS stream
                var subscriptionRequest = new SubscriptionRequest { ClientId = _clientId };
                using var streamingCall = _client.Subscribe(subscriptionRequest, cancellationToken: _cancellationTokenSource.Token);

                Console.WriteLine($"Client {_clientId} subscribed to GDDS stream.");

                // Asynchronously read messages from the stream
                await foreach (var gddsMessage in streamingCall.ResponseStream.ReadAllAsync(_cancellationTokenSource.Token))
                {
                    Console.WriteLine($"Received GDDS: IP={gddsMessage.Ip}, Vehicle={gddsMessage.Veh}, " +
                        $"Lat={gddsMessage.Lat}, Lon={gddsMessage.Lon}, Speed={gddsMessage.Spd}, " +
                        $"Bearing={gddsMessage.Hdg}, TransitMode={(VehicleType)gddsMessage.TrnTyp}, " +
                        $"Timestamp={DateTimeFromUnixMs(gddsMessage.Tm)}, " +
                        $"ServerTimestamp={DateTimeFromUnixMs(gddsMessage.Svrtm)}, Message={gddsMessage.Msg}");
                }
            }
            catch (RpcException ex) when (ex.StatusCode == StatusCode.Cancelled)
            {
                Console.WriteLine($"Client {_clientId} subscription was cancelled.");
            }
            catch (RpcException ex) when (ex.StatusCode == StatusCode.Unauthenticated)
            {
                Console.WriteLine($"Client {_clientId} failed to subscribe: Invalid or unauthorized client ID.");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in subscription for client {_clientId}: {ex.Message}");
            }
        }

        public async Task TerminateAsync()
        {
            if (string.IsNullOrEmpty(_clientId))
            {
                Console.WriteLine("No client ID set. Nothing to terminate.");
                return;
            }

            try
            {
                // Send termination request
                var terminationRequest = new TerminationRequest { ClientId = _clientId };
                var response = await _client.TerminateAsync(terminationRequest);

                if (response.Success)
                {
                    Console.WriteLine($"Client {_clientId} successfully terminated subscription.");
                }
                else
                {
                    Console.WriteLine($"Client {_clientId} termination failed.");
                }

                // Cancel the streaming operation
                _cancellationTokenSource.Cancel();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error terminating client {_clientId}: {ex.Message}");
            }
        }

        private static DateTime DateTimeFromUnixMs(long unixMs)
        {
            return DateTimeOffset.FromUnixTimeMilliseconds(unixMs).UtcDateTime;
        }

        public void Dispose()
        {
            _cancellationTokenSource.Cancel();
            _cancellationTokenSource.Dispose();
        }
    }

    public enum VehicleType
    {
        Unknown = 0,
        Car = 1,
        Truck = 2,
        Bus = 3,
        Motorcycle = 4
    }

    // Example usage
    public class Program
    {
        public static async Task Main()
        {
            string serverAddress = "https://localhost:5001"; // Replace with your server address

            using var client = new GddsClient(serverAddress);

            // Request a client ID from the server
            string clientId = await client.GetClientIdAsync();
            Console.WriteLine($"Received client ID: {clientId}");

            // Start receiving in a background task
            var receiveTask = client.StartReceivingAsync();

            // Simulate some work or wait for user input
            Console.WriteLine("Press any key to terminate the subscription...");
            Console.ReadKey();

            // Terminate the subscription
            await client.TerminateAsync();

            // Wait for the receiving task to complete
            await receiveTask;

            Console.WriteLine("Client shut down.");
        }
    }
}